### PR TITLE
Added State And Replay Mode

### PR DESCRIPTION
**Pull request by SwapnilOutscal**
Added Petrol and Chase State
Added Replay mode to enemy

### CODE REVIEW

File: Assets/Scripts/Enemy/EnemyController.cs [hunk 1]
```diff
@@ -34,12 +41,31 @@ public void TakeDamage(int value)
...
         }
 
+        public void ChangeState(EnemyState enemyState)
+        {
+            lastStateView = currentStateView;
+
+            currentStateView = enemyState;
+
+            if (currentStateView == EnemyState.petrol)
+                enemyView.PetrolState.enabled = true;
```
Comment (mayankgrover) [inline Assets/Scripts/Enemy/EnemyController.cs:69]: you are not disabling the last state?

File: Assets/Scripts/Enemy/StateView/PetrolState.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                wayPointList.Add(wayPoint.GetComponent<Transform>().position);
+            }
+
+            if (GameManager.Instance.currentState.gameStateType == StateMachine.GameStateType.Game)
+            {
+                SelectWayPoint();
+            }
+            else if (GameManager.Instance.currentState.gameStateType == StateMachine.GameStateType.Replay)
+            {
+                currentWayPointIndex = EnemyManager.Instance.EnemyDatas[enemyView.enemyIndex].wayPoints[replayIndexCOunt];
```
Comment (mayankgrover) [inline Assets/Scripts/Enemy/StateView/PetrolState.cs]: the waypoint data needs to be passed into the view from the controller. view should not be going out to services to get data

File: Assets/Scripts/Enemy/StateView/PetrolState.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                EnemyData enemyData = new EnemyData();
+                enemyData.wayPoints = new List<int>();
+                enemyData = EnemyManager.Instance.EnemyDatas[enemyView.enemyIndex];
+                enemyData.wayPoints.Add(r);
+                EnemyManager.Instance.EnemyDatas[enemyView.enemyIndex] = enemyData;
+            }
+        }
+
+        // Update is called once per frame
+        void Update()
```
Comment (mayankgrover) [inline Assets/Scripts/Enemy/StateView/PetrolState.cs]: entire function can be made more readable

**Review by mayankgrover — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Enemy/EnemyController.cs (hunk 1 above)
Comment (SwapnilOutscal) [reply Assets/Scripts/Enemy/EnemyController.cs:69]: Doing it in state itself, when it calls to change the state

**Pull request merged**